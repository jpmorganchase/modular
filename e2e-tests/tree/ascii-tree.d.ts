declare module 'ascii-tree' {
  export function generate(input: string): string;
}
