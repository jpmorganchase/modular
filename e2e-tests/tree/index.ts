import fs from 'fs';
import path from 'path';
import asciiTree from 'ascii-tree';
import hash from '@emotion/hash';

function times(str: string, length: number) {
  return Array.from({ length }, () => str).join('');
}

const defaultIgnores = ['node_modules', '.git', '.DS_Store'];
const defaultHashIgnores = [
  // adding lockfiles here because it can be different
  // on different runs; since we install the latest versions
  // of some packages when making a repository
  'yarn.lock',
  'package-lock.json',
  // adding package.json files since versions within these get bumped
  'package.json',
];

function tree(
  _path: string,
  level = 1,
  options = {
    ignores: defaultIgnores,
    hashIgnores: defaultHashIgnores,
  },
): string {
  const stat = fs.statSync(_path);
  if (stat.isDirectory()) {
    const children = fs.readdirSync(_path);
    const dirArr = _path.split('/');
    const dir = dirArr[dirArr.length - 1];
    // todo - should these be sorted?
    // todo - handle symlinks, etc
    return `${times('#', level)}${dir}\n${children
      .filter((child: string) => !options.ignores.includes(child))
      .map((child: string) => tree(path.join(_path, child), level + 1, options))
      .join('\n')}`;
  } else {
    return [
      `${times('#', level)}${path.basename(_path)}`,
      options.hashIgnores.includes(path.basename(_path))
        ? undefined
        : `#${hash(fs.readFileSync(_path, 'utf8'))}`,
    ]
      .filter(Boolean)
      .join(' ');
  }
}

export default function generate(_path: string): string {
  return asciiTree.generate(tree(_path));
}
