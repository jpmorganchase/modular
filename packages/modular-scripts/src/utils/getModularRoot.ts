import * as path from 'path';
import * as fs from 'fs-extra';
import findUp from 'find-up';

function isModularRoot(packageJson: { modular?: Record<string, unknown> }) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory: string) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

export default function getModularRoot(): string {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
}
