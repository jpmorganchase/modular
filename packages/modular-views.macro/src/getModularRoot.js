// keep this in sync with modular-scripts/src/getModularRoot
const path = require('path');
const fs = require('fs-extra');
const findUp = require('find-up');

function isModularRoot(packageJson) {
  return packageJson?.modular?.type === 'root';
}

function findUpModularRoot() {
  return findUp.sync((directory) => {
    const packageJsonPath = path.join(directory, 'package.json');
    if (
      findUp.sync.exists(packageJsonPath) &&
      isModularRoot(fs.readJsonSync(packageJsonPath))
    ) {
      return packageJsonPath;
    }
    return;
  });
}

module.exports = function getModularRoot() {
  try {
    const modularRoot = findUpModularRoot();
    if (modularRoot === undefined) {
      console.error('These commands must be run within a modular repository.');
      process.exit(1);
    }

    return path.dirname(modularRoot);
  } catch (err) {
    console.error(err);
    return process.exit(1);
  }
};
