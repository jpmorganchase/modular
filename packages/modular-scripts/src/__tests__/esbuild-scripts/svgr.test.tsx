import * as esbuild from 'esbuild';
import * as path from 'path';
import * as fs from 'fs-extra';
import * as tmp from 'tmp';
import tree from 'tree-view-for-tests';
import svgrPlugin from '../../esbuild-scripts/plugins/svgr';
import getModularRoot from '../../utils/getModularRoot';

const emptyDir = async (dirName: string) => {
  await fs.emptyDir(dirName);
};

describe('WHEN running esbuild with the svgrPlugin', () => {
  describe("WHEN there's a url import", () => {
    let tmpDir: tmp.DirResult;
    let result: esbuild.BuildResult;
    let outdir: string;

    beforeAll(async () => {
      tmpDir = tmp.dirSync();
      outdir = path.join(tmpDir.name, 'output');
      result = await esbuild.build({
        entryPoints: [
          path.join(__dirname, '__fixtures__', 'svgr-url', 'index.tsx'),
        ],
        plugins: [svgrPlugin()],
        // so that we don't get huge output sizes...
        external: ['react', 'react-dom'],
        outdir,
        sourceRoot: getModularRoot(),
        bundle: true,
        splitting: true,
        format: 'esm',
        target: 'es2021',
      });
    });

    afterAll(async () => {
      await emptyDir(tmpDir.name);
      tmpDir.removeCallback();
    });

    it('SHOULD be successful', () => {
      expect(result.errors).toEqual([]);
      expect(result.warnings).toEqual([]);
    });

    it('SHOULD have the correct output structure', () => {
      expect(tree(outdir, { hashIgnores: ['index.js'] }))
        .toMatchInlineSnapshot(`
        "output
        ├─ index.js
        └─ logo-5JCTDEME.svg #lzkq0q"
      `);
    });

    it('SHOULD ouput the correct index.js', () => {
      let content = String(fs.readFileSync(path.join(outdir, 'index.js')));
      content = content.replaceAll(getModularRoot(), '');
      expect(content).toMatchSnapshot();
    });

    it('SHOULD ouput the correct logo.svg', () => {
      let content = String(
        fs.readFileSync(path.join(outdir, 'logo-5JCTDEME.svg')),
      );
      content = content.replaceAll(getModularRoot(), '');
      expect(content).toMatchSnapshot();
    });
  });

  describe("WHEN there's a component import", () => {
    let tmpDir: tmp.DirResult;
    let result: esbuild.BuildResult;
    let outdir: string;

    beforeAll(async () => {
      tmpDir = tmp.dirSync();
      outdir = path.join(tmpDir.name, 'output');
      result = await esbuild.build({
        entryPoints: [
          path.join(__dirname, '__fixtures__', 'svgr-component', 'index.tsx'),
        ],
        plugins: [svgrPlugin()],
        // so that we don't get huge output sizes...
        external: ['react', 'react-dom'],
        outdir,
        sourceRoot: getModularRoot(),
        bundle: true,
        splitting: true,
        format: 'esm',
        target: 'es2021',
      });
    });

    afterAll(async () => {
      await emptyDir(tmpDir.name);
      tmpDir.removeCallback();
    });

    it('SHOULD be successful', () => {
      expect(result.errors).toEqual([]);
      expect(result.warnings).toEqual([]);
    });

    it('SHOULD have the correct output structure', () => {
      expect(tree(outdir, { hashIgnores: ['index.js'] }))
        .toMatchInlineSnapshot(`
        "output
        └─ index.js"
      `);
    });

    it('SHOULD ouput the correct index.js', () => {
      let content = String(fs.readFileSync(path.join(outdir, 'index.js')));
      content = content.replaceAll(getModularRoot(), '');
      expect(content).toMatchSnapshot();
    });
  });

  describe("WHEN there's a dataurl import", () => {
    let tmpDir: tmp.DirResult;
    let result: esbuild.BuildResult;
    let outdir: string;

    beforeAll(async () => {
      tmpDir = tmp.dirSync();
      outdir = path.join(tmpDir.name, 'output');
      result = await esbuild.build({
        entryPoints: [
          path.join(__dirname, '__fixtures__', 'svgr-dataurl', 'index.tsx'),
        ],
        plugins: [svgrPlugin()],
        // so that we don't get huge output sizes...
        external: ['react', 'react-dom'],
        outdir,
        sourceRoot: getModularRoot(),
        bundle: true,
        splitting: true,
        format: 'esm',
        target: 'es2021',
      });
    });

    afterAll(async () => {
      await emptyDir(tmpDir.name);
      tmpDir.removeCallback();
    });

    it('SHOULD be successful', () => {
      expect(result.errors).toEqual([]);
      expect(result.warnings).toEqual([]);
    });

    it('SHOULD have the correct output structure', () => {
      expect(tree(outdir, { hashIgnores: ['index.js'] }))
        .toMatchInlineSnapshot(`
        "output
        ├─ index.css #5f8lem
        └─ index.js"
      `);
    });

    it('SHOULD ouput the correct index.js', () => {
      let content = String(fs.readFileSync(path.join(outdir, 'index.js')));
      content = content.replaceAll(getModularRoot(), '');
      expect(content).toMatchSnapshot();
    });

    it('SHOULD ouput the correct index.css', () => {
      let content = String(fs.readFileSync(path.join(outdir, 'index.css')));
      content = content.replaceAll(getModularRoot(), '');
      expect(content).toMatchSnapshot();
    });
  });
});
