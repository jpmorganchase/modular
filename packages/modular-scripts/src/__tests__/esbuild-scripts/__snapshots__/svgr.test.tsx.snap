// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WHEN running esbuild with the svgrPlugin WHEN there's a component import SHOULD ouput the correct index.js 1`] = `
"// packages/modular-scripts/src/__tests__/esbuild-scripts/__fixtures__/svgr-component/logo.svg
import {
  createElement
} from \\"react\\";
function ReactComponent(props) {
  return /* @__PURE__ */ createElement(\\"svg\\", {
    viewBox: \\"0 0 100 100\\",
    xmlns: \\"http://www.w3.org/2000/svg\\",
    ...props
  }, /* @__PURE__ */ createElement(\\"circle\\", {
    cx: 50,
    cy: 50,
    r: 50
  }));
}

// packages/modular-scripts/src/__tests__/esbuild-scripts/__fixtures__/svgr-component/index.tsx
import React2 from \\"react\\";
import ReactDOM from \\"react-dom\\";
ReactDOM.render(/* @__PURE__ */ React2.createElement(ReactComponent, {
  title: \\"logo\\"
}), document.body);
"
`;

exports[`WHEN running esbuild with the svgrPlugin WHEN there's a dataurl import SHOULD ouput the correct index.css 1`] = `
"/* packages/modular-scripts/src/__tests__/esbuild-scripts/__fixtures__/svgr-dataurl/index.css */
.app {
  background-image: url(data:image/svg+xml;base64,PHN2ZyB2aWV3Qm94PSIwIDAgMTAwIDEwMCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KICA8Y2lyY2xlIGN4PSI1MCIgY3k9IjUwIiByPSI1MCIvPgo8L3N2Zz4=);
}
"
`;

exports[`WHEN running esbuild with the svgrPlugin WHEN there's a dataurl import SHOULD ouput the correct index.js 1`] = `
"// packages/modular-scripts/src/__tests__/esbuild-scripts/__fixtures__/svgr-dataurl/index.tsx
import React from \\"react\\";
import ReactDOM from \\"react-dom\\";
ReactDOM.render(/* @__PURE__ */ React.createElement(\\"div\\", {
  className: \\"app\\"
}), document.body);
"
`;

exports[`WHEN running esbuild with the svgrPlugin WHEN there's a url import SHOULD ouput the correct index.js 1`] = `
"// svgurl:packages/modular-scripts/src/__tests__/esbuild-scripts/__fixtures__/svgr-url/logo.svg
var logo_default = \\"./logo-5JCTDEME.svg\\";

// packages/modular-scripts/src/__tests__/esbuild-scripts/__fixtures__/svgr-url/logo.svg
import {
  createElement
} from \\"react\\";

// packages/modular-scripts/src/__tests__/esbuild-scripts/__fixtures__/svgr-url/index.tsx
import React2 from \\"react\\";
import ReactDOM from \\"react-dom\\";
ReactDOM.render(/* @__PURE__ */ React2.createElement(\\"img\\", {
  src: logo_default,
  alt: \\"logo\\"
}), document.body);
"
`;

exports[`WHEN running esbuild with the svgrPlugin WHEN there's a url import SHOULD ouput the correct logo.svg 1`] = `
"<svg viewBox=\\"0 0 100 100\\" xmlns=\\"http://www.w3.org/2000/svg\\">
  <circle cx=\\"50\\" cy=\\"50\\" r=\\"50\\"/>
</svg>"
`;
